synapseclient/core/sts_transfer.py

```diff
@@ -0,0 +1,210 @@
+        self._lock = threading.Lock()
+
+    def get_token(self, syn, entity_id, permission, min_remaining_life: datetime.timedelta = None):
+        min_remaining_life = min_remaining_life if min_remaining_life is not None else self.DEFAULT_MIN_LIFE
+
+        utcnow = datetime.datetime.utcnow()
+        with self._lock:
+            token_cache = self._tokens.get(permission)
+            if token_cache is None:
+                raise ValueError(f"Invalid STS permission {permission}")
+
+            token = token_cache.get(entity_id)
+            if not token or (iso_to_datetime(token['expiration']) - utcnow) < min_remaining_life:
+                # either there is no cached token or the remaining life on the token isn't enough so fetch new
+                token = token_cache[entity_id] = self._fetch_token(syn, entity_id, permission)
+
+        return token
+
+    @staticmethod
+    def _fetch_token(syn, entity_id, permission):
+        return syn.restGET(f'/entity/{entity_id}/sts?permission={permission}')
+
+
+_TOKEN_STORE = _StsTokenStore()
+
+
+def get_sts_credentials(syn, entity_id, permission, output_format='json', **kwargs):
+    """Get STS credentials for the given entity_id and permission, outputting it in the given format
+    """
+
+    value = _TOKEN_STORE.get_token(syn, entity_id, permission, **kwargs)
+    value['secretAccessKey'] = value['secretAccessKey']
+
+    if output_format == 'boto':
+        # the Synapse STS API returns camel cased keys that we need to convert to use with boto.
+        # prefix with "aws_", convert to snake case, and exclude any other key/value pairs in the value
+        # e.g. expiration
+        value = {"aws_{}".format(snake_case(k)): value[k] for k in (
+            'accessKeyId', 'secretAccessKey', 'sessionToken'
+        )}
+
+    elif output_format == 'shell':
+        # make output in the form of commands that will set the credentials into the user's
+        # environment such that they can e.g. run awscli commands
+
+        if platform.system() == 'Windows' and 'bash' not in os.environ.get('SHELL', ''):
+            # if we're running on windows and we can't detect we're running a bash shell
+            # then we make the output compatible for a windows cmd prompt environment.
+            value = f"""\
+setx AWS_ACCESS_KEY_ID "{value['accessKeyId']}"
+setx AWS_SECRET_ACCESS_KEY "{value['secretAccessKey']}"
+setx AWS_SESSION_TOKEN "{value['sessionToken']}"
+"""
+        else:
+            # assume bourne shell compatible (i.e. bash, zsh, etc)
+            value = f"""\
+export AWS_ACCESS_KEY_ID="{value['accessKeyId']}"
+export AWS_SECRET_ACCESS_KEY="{value['secretAccessKey']}"
+export AWS_SESSION_TOKEN="{value['sessionToken']}"
+"""
+    elif output_format != 'json':
+        raise ValueError(f'Unrecognized output_format {output_format}')
+
+    return value
+
+
+def with_boto_sts_credentials(fn, *args, **kwargs):
```

Comment:
I think a context manager doesn't fit here since we're operating on the result of the call (i.e. the exception) not providing wrapping context to the call. A decorator could work but it would need to be pretty proscriptive, in that the wrapped function would have to fit the credentials, and at that point I think it doesn't really become too useful as general decorator since you'd probably have to create a special closure to fit the case anyway.